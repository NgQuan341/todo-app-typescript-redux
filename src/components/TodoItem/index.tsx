import React, { useState } from 'react'
import { Row, Checkbox, Button, Modal, Input, Typography, Divider, Col } from 'antd';
import { useDispatch } from 'react-redux'
// import { removeTodo, updateTodo } from '../../redux/actions';
import todoListSlice, { removeATodo, updateATodo } from '../TodoList/reducer';
import { Item } from '../../model';
import { EditOutlined, CloseOutlined } from '@ant-design/icons'


export interface todoItemProps {
    item: Item
}

const TodoItem = ({ item }: todoItemProps) => {
    const dispatch = useDispatch()
    const [itemUpdate, setItemUpdate] = useState(item)
    const [isModalVisible, setIsModalVisible] = useState(false);
    const [checked, setChecked] = useState(item.completed)

    const onChangeUpdate = (e: React.ChangeEvent<HTMLInputElement>) => { setItemUpdate({ ...itemUpdate, name: e.target.value }) }
    const handleUpdate = (data: Item) => {
        dispatch(todoListSlice.actions.updateTodo(data)) 
        dispatch<any>(updateATodo(data))
    }
    const handleRemove = (data: Item) => {
        dispatch(todoListSlice.actions.removeTodo(data)) 
        dispatch<any>(removeATodo(data))
    }

    const toggleCheckbox = () => {
        setChecked(!checked)
        handleUpdate({ ...itemUpdate, completed: !checked })
    }
    const showModal = () => {
        setIsModalVisible(true);
    };

    const handleOk = () => {
        if (itemUpdate.name !== '') {
            handleUpdate(itemUpdate)
        }
        else {
            setItemUpdate(item)
        }
        setIsModalVisible(false);
    };

    const handleCancel = () => {
        setItemUpdate(item)
        setIsModalVisible(false);
    };
    return (
        <>
            <Row
                justify='space-between'
                style={{
                    marginBottom: 3,
                }}
            >
                <Col span={18} style={{ ...(checked ? { opacity: 0.5, textDecoration: 'line-through' } : {}) }}>
                    <Checkbox checked={checked} onChange={toggleCheckbox} style={{ display: 'flex' }}>
                        {item.name}
                    </Checkbox>
                    <Modal
                        visible={isModalVisible}
                        onOk={handleOk}
                        onCancel={handleCancel}
                    >
                        <Typography>Update Todo</Typography>
                        <Divider />
                        <Input value={itemUpdate.name} onChange={(e) => onChangeUpdate(e)} />
                    </Modal>
                </Col>
                <Col span={4}>
                    <Row justify="space-around">
                        <Button type='primary' icon={<EditOutlined />} onClick={showModal} disabled={checked} 
                        style={{ ...(checked ? { opacity: 0.5, textDecoration: 'line-through' } : {}) }} />
                        <Button type='primary' danger icon={<CloseOutlined />} onClick={() => handleRemove(item)} style={{ opacity: 1 }} />
                    </Row>
                </Col>

            </Row>

        </>
    )
}
export const TypeOfTodoItem = typeof TodoItem

export default TodoItem
