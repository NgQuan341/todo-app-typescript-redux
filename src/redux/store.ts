// import {createStore} from 'redux'
// import rootReducer from './reducer';
// import { composeWithDevTools } from 'redux-devtools-extension'

// const composedEnheancers = composeWithDevTools()
// const store = createStore(rootReducer, composedEnheancers);

// export default store

import { configureStore } from '@reduxjs/toolkit'
import todoListSlice from '../components/TodoList/reducer';

const store = configureStore({
    reducer: todoListSlice.reducer
})
export type RootState = ReturnType<typeof store.getState>

export default store