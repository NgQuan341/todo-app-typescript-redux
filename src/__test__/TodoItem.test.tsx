import { mount, shallow } from 'enzyme'
import { Provider } from 'react-redux';
import TodoItem, { todoItemProps } from '../components/TodoItem';
import store from '../redux/store';

const setup = ({item}:todoItemProps) => {
   let wrapper = shallow<typeof TodoItem>(<Provider store={store}><TodoItem item={item} /></Provider>)
   return wrapper
   
}
describe('TodoItem component', () =>{
    test('renders without error', () => {
        let item = {
            id:'jjjj',
            name:'todo1',
            completed:true
        }
        const wrapper = setup({item}) 
        expect(wrapper.html()).toMatchSnapshot()
    })
})