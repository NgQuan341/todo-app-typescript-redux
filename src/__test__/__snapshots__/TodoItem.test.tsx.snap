// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem component renders without error 1`] = `"<div class=\\"ant-row ant-row-space-between\\" style=\\"margin-bottom:3px\\"><div style=\\"opacity:0.5;text-decoration:line-through\\" class=\\"ant-col ant-col-18\\"><label class=\\"ant-checkbox-wrapper ant-checkbox-wrapper-checked\\" style=\\"display:flex\\"><span class=\\"ant-checkbox ant-checkbox-checked\\"><input type=\\"checkbox\\" class=\\"ant-checkbox-input\\" checked=\\"\\"/><span class=\\"ant-checkbox-inner\\"></span></span><span>todo1</span></label></div><div class=\\"ant-col ant-col-4\\"><div class=\\"ant-row ant-row-space-around\\"><button style=\\"opacity:0.5;text-decoration:line-through\\" type=\\"button\\" class=\\"ant-btn ant-btn-primary ant-btn-icon-only\\" disabled=\\"\\"><span role=\\"img\\" aria-label=\\"edit\\" class=\\"anticon anticon-edit\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"edit\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M257.7 752c2 0 4-.2 6-.5L431.9 722c2-.4 3.9-1.3 5.3-2.8l423.9-423.9a9.96 9.96 0 000-14.1L694.9 114.9c-1.9-1.9-4.4-2.9-7.1-2.9s-5.2 1-7.1 2.9L256.8 538.8c-1.5 1.5-2.4 3.3-2.8 5.3l-29.5 168.2a33.5 33.5 0 009.4 29.8c6.6 6.4 14.9 9.9 23.8 9.9zm67.4-174.4L687.8 215l73.3 73.3-362.7 362.6-88.9 15.7 15.6-89zM880 836H144c-17.7 0-32 14.3-32 32v36c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-36c0-17.7-14.3-32-32-32z\\"></path></svg></span></button><button style=\\"opacity:1\\" type=\\"button\\" class=\\"ant-btn ant-btn-primary ant-btn-icon-only ant-btn-dangerous\\"><span role=\\"img\\" aria-label=\\"close\\" class=\\"anticon anticon-close\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"close\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z\\"></path></svg></span></button></div></div></div>"`;
